### PR TITLE
Turning in Lesson 9 assignments

### PR DESCRIPTION
**Pull request by tfbanks**

### CODE REVIEW

File: students/tfbanks/Lesson09/Mailroom_pt5.py [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+
+donor_data = {
+        'Tim Cooker': [1500.50, 25050.50, 15680.75],
+        'Elon Musket': [5250.25, 26500],
+        'Frank Petersmankempt': [550.60],
+        'Megan Morgan': [650.40, 1600, 20625.40],
+        'Marlene Wheeler': [820, 1222.80]} 
+
+
+class Donor_Actions:
```
Comment (natasha-aleksandrova) [inline students/tfbanks/Lesson09/Mailroom_pt5.py]: classes use UpperCamelCase style: `DonorActions`

File: students/tfbanks/Lesson09/Mailroom_pt5.py [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+# !/usr/bin/env python3
+
+import os
+# Donor Database containing names and donations
+
+
+
+class Donor:
+    def __init__(self, donations=None):
+        self.donations = donations if donations else []
```
Comment (natasha-aleksandrova) [inline students/tfbanks/Lesson09/Mailroom_pt5.py]: your Donor class should have more attributes and methods pertaining to donor object like:
* donor name
* avg donations
* total donations
* sum donations
* thank you letter text

File: students/tfbanks/Lesson09/Mailroom_pt5.py [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+        'Elon Musket': [5250.25, 26500],
+        'Frank Petersmankempt': [550.60],
+        'Megan Morgan': [650.40, 1600, 20625.40],
+        'Marlene Wheeler': [820, 1222.80]} 
+
+
+class Donor_Actions:
+
+    def add_new_donor(donor, donations):
+        donor_data.setdefault((donor),[]).append(donations)
```
Comment (natasha-aleksandrova) [inline students/tfbanks/Lesson09/Mailroom_pt5.py]: your functions should either be class methods 
```
def add_new_donor(self, donor, donations)
```
or staticmethods
```
@staticmethod
def add_new_donor(donor, donations)
```

here it isn't a proper use of class functions.

File: students/tfbanks/Lesson09/Mailroom_pt5.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        donor_data[donor].append(donations)
+
+    #Returns sorted list of all donors
+
+    def all_donors():
+        donors = list(donor_data.keys())
+        return sorted(donors)
+
+    def thank_you_donor(): # inputting a new donor and/or new donations
+        donor = donor_prompt()
```
Comment (natasha-aleksandrova) [inline students/tfbanks/Lesson09/Mailroom_pt5.py]: there should be no `input` function calls in your donor classes; even though actual code that calls input is in outside function it's essentially the same.

**Comment by natasha-aleksandrova** [comment]
I will write up some tips in a minute.

**Comment by natasha-aleksandrova** [comment]
here is a very simple outline:
```
class Donor:
    def __init__(self, name, donations):
        self.name = name
        self.donations = donations

    def add_donation(self, donation):
        self.donations.append(donation)

    @property
    def total_donation(self):
        return len(self.donations)
    
    def get_thank_you_letter(self):
        return f"{self.name} thanks!"


class Donations:
    def __init__(self, *args):
        self.donors = {d.name: d for d in args}

    def add_donation(self, name, donation):
        if self.donors.get(name):
            self.donors[name].add_donation(donation)
        else:
            self.donors[name] = Donor(name, donation)
    
    def get_report(self):
        for donor_obj in self.donors.values():
            donor_obj.total_donation

    def send_letters(self):
        for donor_obj in self.donors.values():
            donor_obj.get_thank_you_letter()


donations = Donations(Donor("natasha", [100]))
donations.add_donation("bob", [40])

```

**Comment by tfbanks** [comment]
Thanks,  I’ll rework it.  I struggled to take the principles from the circle into the mailroom, I started off with a lot of those formulas in the donor class but the action class kept giving me errors but I found what I came up with worked and seemed to follow somewhat.

**Comment by natasha-aleksandrova** [comment]
sounds good 👍 

let me know if you have any questions on the sample code I posted.

Yes going from Circle to Mailroom OO is a step up. Especially implementing relationship between Donor and DonorCollection.

**Pull request merged**